procmonAnalyze.js

```diff
@@ -331,48 +358,72 @@ async function drawData(data, diskify) {
   }
   let diskmapScale = diskmapCanvas.height / maxLcn;
 
+  if (isRefocused) {
+    minTime = gOuterState.constrainedTimeBegin;
+    maxTime = gOuterState.constrainedTimeEnd;      
+  }
+
   let totalTime = maxTime - minTime;
   let trackWidth = canvas.width / tracks.length;
   let rendererScale = canvas.height / totalTime;
-  gState = {
-    minTime,
-    maxTime,
-    tracks,
-    totalTime,
-    trackWidth,
+
+  if (!isRefocused) {
+    gState = {
+      minTime,
+      maxTime,
+      tracks,
+      totalTime,
+      trackWidth,
+      rendererScale,
+      targetRendererScale: rendererScale,
+      diskmapScale,
+      targetDiskmapScale: diskmapScale,
+      readsByPath,
+      diskify,
+      maxLcn,
+      diskmapTranslate: 0,
+      targetDiskmapTranslate: 0,
+      rendererTranslate: 0,
+      targetRendererTranslate: 0,
+      mouseX: 0,
+      mouseY: 0,
+      timelineIndicators: [],
+      lcnReads: [],
+      lastHoveredRect: null,
+      selectedEntry: null,
+      activePath: null,
+    };
+  }
+
+  gOuterState = {
+    constrainedTimeBegin: minTime,
+    constrainedTimeEnd: maxTime,
+    constrainedTotalTime: totalTime,
+    constrainedTimeBeginRelative: absTimeToRelTime(minTime),
+    constrainedTimeEndRelative: absTimeToRelTime(maxTime),
+    topMarkerCoordinate: 0,
+    bottomMarkerCoordinate: markersCanvas.height,
     rendererScale,
-    targetRendererScale: rendererScale,
-    diskmapScale,
-    targetDiskmapScale: diskmapScale,
-    readsByPath,
-    diskify,
-    maxLcn,
-    diskmapTranslate: 0,
-    targetDiskmapTranslate: 0,
-    rendererTranslate: 0,
-    targetRendererTranslate: 0,
-    mouseX: 0,
-    mouseY: 0,
-    timelineIndicators: [],
-    lcnReads: [],
-    lastHoveredRect: null,
-    selectedEntry: null,
-    activePath: null,
+    unconstrainedData: aData,
+    unconstrainedDiskifyData: diskify,
+    constrainedTimeBeginSet: minTime,
+    targetRendererTranslate: gState.minTime-minTime,
   };
 
   renderer.scale(trackWidth, rendererScale);
   renderer.translate(0, 0);
 
   renderer.clearAll();
-  drawBackground();
+  drawBackground(isRefocused);
+  drawRangeMarkers();
   drawForeground();
   renderer.draw();
 
   drawTopPathsInfo();
   scheduleRedrawDiskmap();
 }
 
-function drawBackground() {
+function drawBackground(isRefocused) {
```

Comment:
@squarewave So the problem I'm facing is now with `drawBackground` once we've scrolled. It seems like things are working when we create top/bottom constraints (correct foreground showing, correct timeline indicators, etc). The problem is when we scroll from a specific section. I feel like the problem has something to do with `let offsetPx = (offset + targetRendererTranslate) * rendererScale;`. We would want the timeline indicators to change depending on our scroll, and right now they just push the original numbers up. Do you have any suggestions as to how I can fix this?

Draw foreground seems to be working fine with the scroll, it just has an incorrect background associated with it.